src/System.Reflection.MetadataLoadContext/src/System/Reflection/PathAssemblyResolver.cs

```diff
@@ -0,0 +1,117 @@
+
+        /// <summary>
+        /// Initializes a new instance of the cref="System.Reflection.PathAssemblyResolver"> class.
+        /// </summary>
+        /// <exception cref="System.ArgumentNullException">Thrown when assemblyPaths is null.</exception>
+        /// <exception cref="System.ArgumentException">Thrown when a path is invalid or an Assembly simple name was specified more than once.</exception>
+        public PathAssemblyResolver(IEnumerable<string> assemblyPaths)
+        {
+            if (assemblyPaths == null)
+                throw new ArgumentNullException(nameof(assemblyPaths));
+
+            foreach (string path in assemblyPaths)
+            {
+                if (string.IsNullOrEmpty(path))
+                    throw new ArgumentException(SR.Format(SR.Arg_InvalidPath, path), nameof(assemblyPaths));
+
+                string file = Path.GetFileNameWithoutExtension(path);
+                if (file.Length == 0)
+                    throw new ArgumentException(SR.Format(SR.Arg_InvalidPath, path), nameof(assemblyPaths));
+
+                List<string> paths;
+                if (!_fileToPaths.TryGetValue(file, out paths))
+                {
+                    _fileToPaths.TryAdd(file, paths = new List<string>());
+                }
+                paths.Add(path);
+            }
+        }
+
+        public override Assembly Resolve(MetadataLoadContext context, AssemblyName assemblyName)
+        {
+            // On the first access to this assembly name, load each assembly from the provided path(s).
+            LoadMatchingAssembliesFromPaths(context, assemblyName.Name);
+
+            Assembly candidate = null;
+
+            foreach (Assembly a in context.GetAssemblies())
+            {
+                if (assemblyName.Name.Equals(a.GetName().Name, StringComparison.OrdinalIgnoreCase) && IsPublicKeyTokenCompatible(assemblyName, a))
+                {
+                    // Pick the highest version.
+                    if (candidate == null || NormalizeVersion(a.GetName().Version) > NormalizeVersion(candidate.GetName().Version))
+                    {
+                        candidate = a;
+                    }
+                }
+            }
+
+            return candidate;
+        }
+
+        private void LoadMatchingAssembliesFromPaths(MetadataLoadContext context, string simpleName)
+        {
+            List<string> paths;
+            if (_fileToPaths.TryGetValue(simpleName, out paths))
+            {
+                foreach (string path in paths)
+                {
+                    context.LoadFromAssemblyPath(path);
+                }
+
+                // Remove the path so we don't process it again.
+                _fileToPaths.TryRemove(simpleName, out paths);
+            }
+        }
+
+        private bool IsPublicKeyTokenCompatible(AssemblyName name, Assembly a)
+        {
+            if (name.GetPublicKeyToken() == null)
```

Comment:
I had to check that the repeated calls to GetPublicKeyToken weren't each allocating fresh arrays. It's common for methods returning arrays to do so. To save future readers from having to check, I recommend putting the arrays into locals.

Reply:
Ok - I'll add some local vars